File: packages/runtime-urhox-lua/adapter/PageFrame.lua
local PageFrame = {}

function PageFrame.Calculate(viewportWidth, viewportHeight, designWidth, designHeight, left, top, right, bottom)
    if not designWidth or designWidth <= 0 or not designHeight or designHeight <= 0 then error("LUI 页面设计尺寸必须为正数。") end
    left, top, right, bottom = left or 0, top or 0, right or 0, bottom or 0
    local availableWidth = math.max(0, viewportWidth - left - right)
    local availableHeight = math.max(0, viewportHeight - top - bottom)
    local scale = math.min(availableWidth / designWidth, availableHeight / designHeight)
    if scale ~= scale or scale == math.huge then scale = 1 end
    scale = math.max(0, scale)
    local width, height = designWidth * scale, designHeight * scale
    return { x = left + (availableWidth - width) * 0.5, y = top + (availableHeight - height) * 0.5, width = width, height = height, scale = scale, availableWidth = availableWidth, availableHeight = availableHeight }
end

return PageFrame

File: packages/spec/src/page-frame.ts
export interface PageFrameInput { viewportWidth: number; viewportHeight: number; designWidth: number; designHeight: number; marginLeft?: number; marginTop?: number; marginRight?: number; marginBottom?: number; }
export interface PageFrame { x: number; y: number; width: number; height: number; scale: number; availableWidth: number; availableHeight: number; }

export function calculatePageFrame(input: PageFrameInput): PageFrame {
  const left = input.marginLeft ?? 0; const top = input.marginTop ?? 0; const right = input.marginRight ?? 0; const bottom = input.marginBottom ?? 0;
  if (!(input.designWidth > 0) || !(input.designHeight > 0)) throw new Error("LUI 页面设计尺寸必须为正数。");
  const availableWidth = Math.max(0, input.viewportWidth - left - right);
  const availableHeight = Math.max(0, input.viewportHeight - top - bottom);
  const rawScale = Math.min(availableWidth / input.designWidth, availableHeight / input.designHeight);
  const scale = Number.isFinite(rawScale) ? Math.max(0, rawScale) : 1;
  const width = input.designWidth * scale; const height = input.designHeight * scale;
  return { x: left + (availableWidth - width) * .5, y: top + (availableHeight - height) * .5, width, height, scale, availableWidth, availableHeight };
}
